CDAP-11391 Upload doesn't require workspace as pre-requisite. If the …

nitinmotgi opened the pull request:
…workspace doesn't exist, then API will automatically create it

cdap-apps committed:
CDAP-11391 Upload doesn't require workspace as pre-requisite. If the workspace doesn't exist, then API will automatically create it

cdap-apps committed:
CDAP-11391 Upload doesn't require workspace as pre-requisite. If the workspace doesn't exist, then API will automatically create it

cdap-apps committed:
CDAP-11391 Upload doesn't require workspace as pre-requisite. If the workspace doesn't exist, then API will automatically create it

russorat commented on service/src/main/java/co/cask/wrangler/dataset/workspace/WorkspaceDataset.java:
```diff
@@ -128,6 +128,28 @@ public void createWorkspaceMeta(String id, String name, DataType type,
+  public boolean hasWorkspace(String id) throws WorkspaceException {
+    try {
+      Row row = table.get(toKey(id));
+      if (row != null) {
```
table.get never returns null. I think you need to check isEmpty() or something.

nitinmotgi replied:
Addressed comment -- not sure why it worked. But using isEmpty is right way to do it.

cdap-apps committed:
CDAP-11391 Addressed PR comments

awholegunch commented on service/src/main/java/co/cask/wrangler/dataset/workspace/WorkspaceDataset.java:
```diff
@@ -128,6 +128,28 @@ public void createWorkspaceMeta(String id, String name, DataType type,
+      return false;
+    } catch (DataSetException e) {
+      throw new WorkspaceException(
+        String.format("Unable to check if workspace exists '%s'",
```
What about using:
```
String.format("Unable to check if workspace '%s' exists '%s'", id,

awholegunch commented on service/src/main/java/co/cask/wrangler/service/directive/DirectivesService.java:
```diff
@@ -265,6 +267,106 @@ public void get(HttpServiceRequest request, HttpServiceResponder responder,
+
+        default: {
+          error(responder, "Invalid content type. Supports text/plain, application/octet-stream " +
+            "and application/data-prep");
```
What about:
```
          error(responder, "Invalid content type. Supports text/plain, application/octet-stream, " +
             "or application/data-prep.");

awholegunch commented on service/src/main/java/co/cask/wrangler/service/directive/DirectivesService.java:
```diff
@@ -265,6 +267,106 @@ public void get(HttpServiceRequest request, HttpServiceResponder responder,
+      // Extract content.
+      byte[] content = handler.getContent();
+      if (content == null) {
+        error(responder, "Body not present, please post the file containing the records to be wrangle.");
```
What about:
```
"Body not present; missing a file containing records to be processed."

elfenheart commented on service/src/main/java/co/cask/wrangler/service/directive/DirectivesService.java:
```diff
@@ -265,6 +267,106 @@ public void get(HttpServiceRequest request, HttpServiceResponder responder,
+
+    try {
+      String name = request.getHeader(PropertyIds.FILE_NAME);
+      String id = ServiceUtils.generateMD5(request.getHeader(name));
```
should this be `ServiceUtils.generateMD5(name)`?

elfenheart commented on service/src/main/java/co/cask/wrangler/service/directive/DirectivesService.java:
```diff
@@ -265,6 +267,106 @@ public void get(HttpServiceRequest request, HttpServiceResponder responder,
+      // if workspace doesn't exist, then we create the workspace before
+      // adding data to the workspace.
+      if (!table.hasWorkspace(id)) {
+        table.createWorkspaceMeta(id, id);
```
can the displayName of the workspace be the file name?

elfenheart replied:
And also perhaps the first column should be `body` instead of the workspace id as well

Commit:
- Fixed a bug in deriving the workspace name
- Change display name of the workspace to the filename
- Name the first column 'body'

bdmogal commented:
@nitinmotgi @elfenheart I've pushed a couple of commits. please review and merge if it looks ok.

elfenheart reviewed (approved):
Tested the change, and it works. 👍  LGTM
